### ISSUE TITLE
Comment parsing problem in ternjs_0.10.0

### ISSUE THREAD

**Issue by sieira** [marijnh/tern#594]
Tern is failing to parse js when adding multiline comments within a statement

This works :

    /**
      * Angular module
      */
    (function() {
        angular.module("angular-test",[])
        .controller('TestController', ['$scope', function($scope) {
            $scope.test = 'Hola!';
        }]);
    })();

This doesn't:

    /**
     * Angular module
     */
    (function() {
        angular.module("angular-test",[])
        /**
         * This comment will make tern explode
         */
        .controller('TestController', ['$scope', function($scope) {
            $scope.test = 'Hola!';
        }]);
    })();

As many angular libraries use this commenting style, this is making eclipse angular dev tools useless. Please note that I'm not sure about the origin of this bug, I tracked it down from an unexpected behaviour in angularjs-eclipse.

You can take a look to [this issue] (https://github.com/angelozerr/angularjs-eclipse/issues/155) for further information

**The exception being thrown is:**
> /plugins/ternjs_0.10.0.201506172156/node_modules/tern/lib/infer.js:301
      this.construct.apply(this, arguments);
                     ^
TypeError: Cannot call method 'apply' of undefined
    at Object.ctor (/home/luis/.p2/pool/plugins/ternjs_0.10.0.201506172156/node_modules/tern/lib/infer.js:301:22)
    at Injector.set (/home/luis/.p2/pool/plugins/ternjs_0.10.0.201506172156/node_modules/tern/plugin/angular.js:37:49)
    at angular_regFieldCall (/home/luis/.p2/pool/plugins/ternjs_0.10.0.201506172156/node_modules/tern/plugin/angular.js:166:22)
    at /home/luis/.p2/pool/plugins/ternjs_0.10.0.201506172156/node_modules/tern/plugin/angular.js:106:5
    at fn.computeRet (/home/luis/.p2/pool/plugins/ternjs_0.10.0.201506172156/node_modules/tern/lib/def.js:241:21)
    at Object.exports.IsCallee.constraint.addType (/home/luis/.p2/pool/plugins/ternjs_0.10.0.201506172156/node_modules/tern/lib/infer.js:368:9)
    at withWorklist (/home/luis/.p2/pool/plugins/ternjs_0.10.0.201506172156/node_modules/tern/lib/infer.js:799:21)
    at Object.extend.propagate (/home/luis/.p2/pool/plugins/ternjs_0.10.0.201506172156/node_modules/tern/lib/infer.js:91:25)
    at inferExprVisitor.CallExpression (/home/luis/.p2/pool/plugins/ternjs_0.10.0.201506172156/node_modules/tern/lib/infer.js:1166:14)
    at Object.CallExpression (/home/luis/.p2/pool/plugins/ternjs_0.10.0.201506172156/node_modules/tern/lib/infer.js:1017:7)

**Comment by marijnh** [comment]
I couldn't get anything bad to happen when I paste this into Tern's demo (http://ternjs.net/doc/demo.html) or try to run it on that file locally with the `angular` plugin enabled. Maybe you can define 'explode' a little more precisely and create a test case that doesn't involve Eclipse?

**Comment by sieira** [comment]
Well, I would love to do it, but I don't know how. I can start a server from my node project, but I don't have any idea about how can I connect to it.

This is the working request/response flow in angularjs-eclipse (the code snippet without comments)

    -----------------------------------
    Tern request#: 
    Server started at 35311 in 232ms
    {
        "files":
        [{
            "name":"public/js/app.js"
           ,"text":"/**\n *\n * Angular module\n */\n(function() {\n\tangular.module(\"angular-test\",[])\n\t.controller('TestController', ['$scope', function($scope) {\n\t  $scope.test = 'Hola!';\n\t}]);\n})();"
           ,"type":"full"
        }]
    }
    Tern response# with 18ms: 
    {}
    -----------------------------------

This is the one failing:

    -----------------------------------
    Tern request#angular_completions: 
    {
        "query":
        {
            "type":"angular"
           ,"subtype":"completions"
           ,"angularTypes":["module"]
           ,"expression":""
           ,"files":["public/js/app.js"]}}
    -----------------------------------
    Tern request#: 
    {
        "files":
        [{
            "name":"public/js/app.js"
           ,"text":"/**\n *\n * Angular module\n */\n(function() {\n\tangular.module(\"angular-test\",[])\n   /**\n    * This will make tern explode\n    */\n\t.controller('TestController', ['$scope', function($scope) {\n\t  $scope.test = 'Hola!';\n\t}]);\n})();"
           ,"type":"full"
        }]
    }

    Tern response# with 13ms: 
    {}
    -----------------------------------

So looks like things are OK, but then:

>infer.js:301
      this.construct.apply(this, arguments);
                     ^
>TypeError: Cannot call method 'apply' of undefined
    at Object.ctor (/home/luis/.p2/pool/plugins/ternjs_0.10.0.201506172156/node_modules/tern/lib/infer.js:301:22)
    at Injector.set (/home/luis/.p2/pool/plugins/ternjs_0.10.0.201506172156/node_modules/tern/plugin/angular.js:37:49)
    at angular_regFieldCall (/home/luis/.p2/pool/plugins/ternjs_0.10.0.201506172156/node_modules/tern/plugin/angular.js:166:22)
    at /home/luis/.p2/pool/plugins/ternjs_0.10.0.201506172156/node_modules/tern/plugin/angular.js:106:5
    at Object.exports.IsCallee.constraint.addType (/home/luis/.p2/pool/plugins/ternjs_0.10.0.201506172156/node_modules/tern/lib/infer.js:368:9)
    at withWorklist (/home/luis/.p2/pool/plugins/ternjs_0.10.0.201506172156/node_modules/tern/lib/infer.js:799:21)
    at fn.computeRet (/home/luis/.p2/pool/plugins/ternjs_0.10.0.201506172156/node_modules/tern/lib/def.js:241:21)
    at Object.extend.propagate (/home/luis/.p2/pool/plugins/ternjs_0.10.0.201506172156/node_modules/tern/lib/infer.js:91:25)
    at inferExprVisitor.CallExpression (/home/luis/.p2/pool/plugins/ternjs_0.10.0.201506172156/node_modules/tern/lib/infer.js:1166:14)
    at Object.CallExpression (/home/luis/.p2/pool/plugins/ternjs_0.10.0.201506172156/node_modules/tern/lib/infer.js:1017:7)

No extra request is being traced in the working example, so looks like the problem is somewhere within the "text" property of "query", as I said, I'd like to provide a test case, but I'm sorry I don't know how.

As for the request/response log I can see from tern console, I wouldn't say it's related to eclipse or angularjs-eclipse, but I don't have any way to be sure.

My project is launching tern server through Angelo zerr's tern-eclipse / angularjs-eclipse this way :

    Nodejs Commnand: "/usr/local/bin/node" "/home/luis/.p2/pool/plugins/ternjs_0.10.0.201506172156/node_modules/tern/bin/tern" --no-port-file --disable-loading-local
    Project dir: /home/luis/workspace/AngularJS-test
    .tern-project: {"ide":{"scriptPaths":[{"type":"FOLDER","path":"public"}]},"plugins":{"guess-types":{},"node":{},"angular":{},"node-express":{},"node-mongoose":{}},"libs":["ecma5","browser"]}

**Comment by angelozerr** [comment]
I will see this problem next month and give you feedback.

**Comment by marijnh** [comment]
The easiest way to reproduce crashes is usually the `bin/condense` tool (which will analyze some code and try to give you a json representation of the types). It takes command-line arguments to load plugins and definitions.

**Comment by sieira** [comment]
I get the same thing from the console, but that's just for tern@0.11.1

@0.12.0 and @0.13.0 are working fine, so looks like it's up to @angelozerr, thanks for your time

    ~/.p2/pool/plugins/ternjs_0.10.0.201506172156/node_modules/tern/bin$ ./condense ~/workspace/AngularJS-test/public/js/app.js --plugin angular

    ~/.p2/pool/plugins/ternjs_0.10.0.201506172156/node_modules/tern/lib/infer.js:301
      this.construct.apply(this, arguments);
                     ^
    TypeError: Cannot call method 'apply' of undefined
    at Object.ctor (~/.p2/pool/plugins/ternjs_0.10.0.201506172156/node_modules/tern/lib/infer.js:301:22)
    at Injector.set (~/.p2/pool/plugins/ternjs_0.10.0.201506172156/node_modules/tern/plugin/angular.js:37:49)
    at angular_regFieldCall (~/.p2/pool/plugins/ternjs_0.10.0.201506172156/node_modules/tern/plugin/angular.js:166:22)
    at ~/.p2/pool/plugins/ternjs_0.10.0.201506172156/node_modules/tern/plugin/angular.js:106:5
    at fn.computeRet (~/.p2/pool/plugins/ternjs_0.10.0.201506172156/node_modules/tern/lib/def.js:241:21)
    at Object.exports.IsCallee.constraint.addType (/home/luis/.p2/pool/plugins/ternjs_0.10.0.201506172156/node_modules/tern/lib/infer.js:368:9)
    at withWorklist (~/.p2/pool/plugins/ternjs_0.10.0.201506172156/node_modules/tern/lib/infer.js:799:21)
    at Object.extend.propagate (~/.p2/pool/plugins/ternjs_0.10.0.201506172156/node_modules/tern/lib/infer.js:91:25)
    at inferExprVisitor.CallExpression (~/.p2/pool/plugins/ternjs_0.10.0.201506172156/node_modules/tern/lib/infer.js:1166:14)
    at Object.CallExpression (~/.p2/pool/plugins/ternjs_0.10.0.201506172156/node_modules/tern/lib/infer.js:1017:7)

**Comment by angelozerr** [comment]
I have fixed this error inside tern.java. It was my fault. As I have a fork of angular.js, I have forgotten to change SeDoc https://github.com/marijnh/tern/commit/7eb972207af545a2224c939f68cbeaac36739e24#diff-ffad115ba87519a141dc5c55a7590a62R11

Problem is fixed inside tern.java. Sorry for the noise @marijnh

**Comment by herrminni** [comment]
Hi Angelo,
thank you for the quick fix! 
I can confirm that your bugfix is working in Eclipse with the SNAPSHOT update-site: 
http://www.nodeclipse.org/updates/anide/SNAPSHOT/

**Comment by angelozerr** [comment]
@herrminni thanks for your feedback. As I have forked tern angular, there were this problem. So I have decided to create https://github.com/angelozerr/tern-angularjs with my fork and I'm writing a lot of tests. 

I would like to contribute to the official tern angular, but as my fork have a lot of changes, I have decided to contribute step by step. Here my first PR https://github.com/marijnh/tern/pull/604
